src/swarm-lang/Swarm/Language/Context.hs

```diff
@@ -87,18 +219,85 @@ vars = M.keys . unCtx
 
 -- | Add a key-value binding to a context (overwriting the old one if
 --   the key is already present).
-addBinding :: Var -> t -> Ctx t -> Ctx t
-addBinding x t (Ctx c) = Ctx (M.insert x t c)
+addBinding :: Hashable t => Var -> t -> Ctx t -> Ctx t
+addBinding x t (Ctx m s@(CtxTree h _)) = Ctx (M.insert x t m) s'
+ where
+  s' = CtxTree h' (CtxUnion (CtxTree tHash (CtxSingle x t)) s)
+  tHash = singletonHash x t
+  h' = case M.lookup x m of
+    Nothing -> h + tHash
+    Just t' -> h - singletonHash x t' + tHash
```

Comment:
These hashes are spread out, could they be written in one function taking restructured tree to `Const Hash`? 🤔

Reply:
Hmm, good question.  When I first saw your comment I thought "oh, of course, I will make a function like `hashTree :: Hashable t => CtxF CtxTree t -> CtxTree t` which rolls up one level of recursion while computing the appropriate top-level hash."  Unfortunately I quickly found that function cannot be written efficiently --- the problem is that in the delete and union cases, we also need the `Map Var t`'s in order to compute the hash (since we need to subtract off overlap when unioning, etc.).

Hmm, maybe I can write something like `Hashable t => CtxF Ctx t -> Ctx t`?  Let me try that...

Reply:
That worked great!  All hash computations are now in a single place, and all other definitions are one-liners.  Thanks for the suggestion!